src/controllers/aer_controller.hpp

```diff
@@ -1042,809 +636,128 @@ Result Controller::execute(std::vector<Circuit> &circuits,
-          circ, noise, config, Method::tensor_network, result);
-    }
-  }
-  default:
-    throw std::runtime_error("Controller:Invalid simulation method");
-  }
-}
-
-//-------------------------------------------------------------------------
-// Utility methods
-//-------------------------------------------------------------------------
-
-size_t Controller::required_memory_mb(const Circuit &circ,
-                                      const Noise::NoiseModel &noise,
-                                      const Method method) const {
-  switch (method) {
-  case Method::statevector: {
-    if (sim_precision_ == Precision::Single) {
-      Statevector::State<QV::QubitVector<float>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
-    } else {
-      Statevector::State<QV::QubitVector<double>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
-    }
-  }
-  case Method::density_matrix: {
-    if (sim_precision_ == Precision::Single) {
-      DensityMatrix::State<QV::DensityMatrix<float>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
-    } else {
-      DensityMatrix::State<QV::DensityMatrix<double>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
-    }
-  }
-  case Method::unitary: {
-    if (sim_precision_ == Precision::Single) {
-      QubitUnitary::State<QV::UnitaryMatrix<float>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
+      return std::make_shared<CircuitExecutor::Executor<
+          QubitSuperoperator::State<QV::Superoperator<double>>>>();
     } else {
-      QubitUnitary::State<QV::UnitaryMatrix<double>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
+      return std::make_shared<CircuitExecutor::Executor<
+          QubitSuperoperator::State<QV::Superoperator<float>>>>();
     }
-  }
-  case Method::superop: {
-    if (sim_precision_ == Precision::Single) {
-      QubitSuperoperator::State<QV::Superoperator<float>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
-    } else {
-      QubitSuperoperator::State<QV::Superoperator<double>> state;
-      return state.required_memory_mb(circ.num_qubits, circ.ops);
-    }
-  }
+    break;
   case Method::stabilizer: {
-    Stabilizer::State state;
-    return state.required_memory_mb(circ.num_qubits, circ.ops);
-  }
+    return std::make_shared<CircuitExecutor::Executor<Stabilizer::State>>();
+  } break;
   case Method::extended_stabilizer: {
-    ExtendedStabilizer::State state;
-    return state.required_memory_mb(circ.num_qubits, circ.ops);
-  }
+    return std::make_shared<
+        CircuitExecutor::Executor<ExtendedStabilizer::State>>();
+  } break;
   case Method::matrix_product_state: {
-    MatrixProductState::State state;
-    return state.required_memory_mb(circ.num_qubits, circ.ops);
-  }
+    return std::make_shared<
+        CircuitExecutor::Executor<MatrixProductState::State>>();
```

Comment:
Why not `MultiStateExecutor` is used for MPS and others?

Reply:
Only a simulation method who supports shot-branching inherits `MultiStateExecutor` at this time.